### ISSUE TITLE
Whether the build option in the docker-compose.yml should be migrated to docker-compose.dev.yml

### ISSUE THREAD

**Issue by cobolbaby** [linkedin/datahub#1825]
As far as I know, the docker/quickstart.sh only start the service and will not involve the compilation process, but dev.sh will execute the build compilation. So whether to migrate the build options to the docker-compose.dev.yml.

**Comment by jplaisted** [comment]
Quickstart is just that; it is meant to people wanting to quickly start or otherwise evaluate DataHub without the need to build anything (all images pulled from DockerHub).

Dev does a mix of building and pulling images. Dependencies (kafka, zookeeper, etc) are still pulled. DataHub related images are ideally[1] "built" locally. However, these images are _extremely_ minimal; they don't build the source code. The dev.yml file mounts your locally built jars to those images instead. The idea is a faster dev flow since building locally will be faster than building on Docker itself. This should be covered [here](https://github.com/linkedin/datahub/blob/master/docs/docker/development.md).

To actually, really, truly rebuild images, as if we were about to publish them, run `COMPOSE_DOCKER_CLI_BUILD=1 DOCKER_BUILDKIT=1 docker-compose -p datahub build
` (as documented [here](https://github.com/linkedin/datahub/tree/master/docker#building-images)). That will take quite awhile though :)


tl;dr quickstart.sh - "I want the latest version of DataHub run asap". dev.sh - "I want my locally built code running asap". Normal docker flow (`build`) - "I want to rebuild the published images from scratch".


Closing for now. Feel free to reopen if this didn't answer your question.


[1] not supported by frontend yet. Only gms and the mce/mae jobs. See #1807.

**Comment by cobolbaby** [comment]
@jplaisted , I understand your thoughts, but if you put the build option in the docker-compose.yml file, it will build a new image instead of pulling the image directly from the docker hub. This is obviously not the expected result.

Ref: https://docs.docker.com/compose/compose-file/#build

> If you specify image as well as build, then Compose names the built image with the webapp and optional tag specified in image

**Comment by cobolbaby** [comment]
@mars-lan Could you reopen the issue.

**Comment by mars-lan** [comment]
Thanks for the doc, @cobolbaby. I don't think there's anything wrong here specifying `build` in [`docker-compose.yml`](https://github.com/linkedin/datahub/blob/master/docker/docker-compose.yml). What the doc says is that **if** you run `docker-compose --build` command, it'll create a local image with the name & tag specified in `image` field, which is what https://github.com/linkedin/datahub/tree/master/docker#building-images is about. For quickstart, we simply run `docker-compose pull &&  docker-compose up` so it will pull the latest (or specific version of) images from docker hub first, then use them to start, **without** any rebuilding any local images.

**Comment by jplaisted** [comment]
Right, what Mars said :) `docker-compose up` will only build images that are "missing" by default. But if you run `docker-compose pull` first (which quickstart.sh does!) then it won't build anything. You can explicitly build with `docker-compose build` or `docker-compose up --build`.


Also notice that the build instructions for `docker-compose.yml` and `docker-compose.dev.yml` are technically different!

**Comment by jplaisted** [comment]
And to reemphasize, you should really be using one of our two scripts (`dev.sh` or `quickstart.sh`). There shouldn't be much of a reason to want to do `docker-compose build` manually, and it'll take forever.

**Comment by cobolbaby** [comment]
Got it, Thanks for your answers @jplaisted @mars-lan
